### ISSUE TITLE
stft: time dimension might be off

### ISSUE THREAD

**Issue by OverLordGoldDragon** [librosa/librosa#1278]
If the idea is to center the window at each point of the unpadded input, starting at sample 0, and incremented by `hop_length`, then `librosa` produces one excess sample for some combinations of even/odd `len(x)`, `n_fft`, & `win_len`. I've verified the formula to be `(len(x) - 1)//hop_length + 1`; `librosa`'s disagreements with it can be found [here](https://github.com/OverLordGoldDragon/ssqueezepy/blob/gmw/tests/reconstruction_test.py#L189).

If this _is_ intentional, what's the rationale?

**Comment by bmcfee** [comment]
Please use the issue template, and provide a minimal example demonstrating the expected and actual behavior in question.

**Comment by OverLordGoldDragon** [comment]
```python
import numpy as np
from librosa import stft


for N in (512, 513):
  for hop_len in (1, 2, 3):
    for n_fft in (512, 513):
      for win_len in (N//8, N//8 - 1):
         x = np.random.randn(N)
         lSx = stft(x, n_fft=n_fft, hop_length=hop_len, win_length=win_len)

         timeshifts = lSx.shape[-1]
         assert timeshifts == (len(x) - 1) // hop_len
```

**Comment by bmcfee** [comment]
> Please use the issue template, and provide a minimal example **demonstrating the expected and actual behavior** in question.

**Comment by OverLordGoldDragon** [comment]
Above demonstrates `librosa` failing to produce `Sx.shape[1] == (len(x) - 1) // hop_len + 1`, so it's about verifying that formula. Example:

```python
win_length, n_fft, len(x) == 5, 12, 12
[1  2  3  4  5  6  7  8  9  10 11 12] == x
[4  3  2  1  2  3  4  5  6  7  8  9  10 11 12 11 10 9] == x_padded
   [2  3  4  3  2]
            [2  3  4  3  2]
                     [2  3  4  3  2]
                              [2  3  4  3  2]
```
So we expect **four** windows centered within the **original** unpadded signal. Demo:

```python
import numpy as np
import librosa
import ssqueezepy

x = np.random.randn(12)
win_len, n_fft, hop_len = 5, 12, 3

lSx = librosa.stft(   x, win_length=win_len, n_fft=n_fft, hop_length=hop_len)
sSx = ssqueezepy.stft(x, win_len=win_len,    n_fft=n_fft, hop_len=hop_len)
print(lSx.shape, sSx.shape)
```
```python
>>> (7, 5) (7, 4)
```

**Comment by bmcfee** [comment]
Here is the exact sequence of steps that will happen here:

1. x will be padded by `n_fft // 2` samples on either side.  Call the result `x_pad`.  (see `core.stft`).
2. `x_pad` will be framed using `hop_len` and `n_fft` as framing parameters
3. framing will produce `n_frames = 1 + (len(x_pad) - frame_length) // hop_length` frames (see `util.frame`)

In your example, we have the following:

- `len(x) == 12`
- `n_fft == 12`
- `len(x_pad) == 12 + 2 * (12//2) == 24`
- `hop_length == 3`
- `n_frames == 1 + (24 - 12) // 3 == 1  + 12//3  = 5`

So I think this is all working as expected?

**Comment by OverLordGoldDragon** [comment]
The issue is having different rationale for framing; under my scheme, the idea is to **center `window` at each sample of the original sequence `x`, starting at sample 0, strided by `hop_length`**. This is shown in my example above. Librosa outputs _one extra_, meaning the window is necessarily centered _outside of `x`_.

The difference is no gamechanger by any means, but I wonder what librosa's rationale is for planting a window outside the original sequence, which does incur more boundary effects than any window inside the sequence. Put differently, what are the "starting and stopping criterion"? One could go further and include every window whose nonzero support touches `x`, for example. Mine's in bold.

**Comment by bmcfee** [comment]
Ah, I see what you mean now.  I think this all comes down to the amount of left- and right-padding done inside the STFT, and not the framing function per se.

> I wonder what librosa's rationale is for planting a window outside the original sequence, which does incur more boundary effects than any window inside the sequence.

It's been quite some time since we first implemented this (01-31-2014!), so my memory may be hazy.  It might be just as simple as "left-padding by half a frame is correct, so we may as well right-pad by the same amount".  Once the padding amount is determined, the framing operation just does whatever makes sense under the assumption that a full frame is valid.

**Comment by OverLordGoldDragon** [comment]
I see. Indeed the framing logic is identical between `librosa` and `ssqueezepy`, and the only difference's in padding; I've designed it specifically to achieve the described effect. If interested, the logic's contained [here](https://github.com/OverLordGoldDragon/ssqueezepy/blob/0.6.0/ssqueezepy/_stft.py#L129) - summarizing:

 1. Pad `x` to total length of `padlength = len(x) + n_fft - 1`
 2. Pad from left / right by `n1` / `n2`, respectively, where:

```python
if abs(padlength - N) % 2 == 0:  # N == len(x)
    n1 = n2 = (padlength - N) // 2
else:
    n2 = (padlength - N) // 2
    n1 = n2 + 1
```
Rationale is, if we must pad by different amounts from left & right, we pad from left more than from right, because that can achieve leakless padding with `'reflect'` (or simply better) on a pure tone. Suppose `x = np.cos(2*np.pi * np.linspace(0, 1, 128))`; `'reflect'` will insert `x[-2]` to the right of `x[-1]`, so to perfectly continue the sinusoid, we must use `endpoint=True`. Then, if `N` is `1` + power of 2, e.g. `129`, we get a perfect DFT with just one non-zero sample. If `N` isn't such, we still get less leakage.

Left > right because the perfect case will have `x_padded[0] == x[0]` and `x_padded[-1] == x[-2]`, and effect carries over to imperfect.

**Comment by bmcfee** [comment]
Yeah, this all makes sense.  I guess if we wanted to make this really precise, we'd minimize `n_right` such that 

```python
(len(x) + n_fft // 2 + n_right - frame_length) % hop_length == 0
```
where the left-padding is `n_fft // 2` and `frame_length = n_fft`.  This way we get a final frame that has the fewest number of padded samples.

This does potentially raise a broader question about stft and padding, which is somewhat independent of centering.  Should we support "modes" like convolution does?  (Full, valid, the different flavors of same, etc.)  I've never seen this provided by an stft implementation, but I guess I could imagine use cases where one might want more control over the different padding options.  Should we include frames that touch any sample (full), or only those which contain no padding (ie, `center=False` currently)?  Our `center=True` is somewhere in between, as apparently is ssqueezepy's if I understand correctly.  It may be worth making these various options more explicit, or at least better documented.

@lostanlen do you have any opinions on this?

**Comment by OverLordGoldDragon** [comment]
Indeed a convolution-like padding support is viable. I'll draw consideration toward one scheme in particular: zero-padding. The problem with most padding schemes is imposing some assumption on boundary behaviors, which real signals may or may not obey. Non-compliants can incur significant effects, like splitting a ridge in two, especially toward lower frequencies. 

Strangely, the most naive scheme also seems most robust to all sorts of boundary behaviors: zeros. While nothing is remedied, nothing is _made worse_ either, which in some cases is certainly a net-plus, and thus a worthwhile user option. We get performance as a bonus, as we need not actually do any padding, just multiply `window` by frames correctly, and maybe take longer FFT.

<details>
  <summary>Demo code</summary>

```python
import numpy as np
from ssqueezepy import ssq_stft
from ssqueezepy.visuals import imshow

t = np.linspace(0, 1, 256, 1)
x = np.cos(2*np.pi * 16 * t)

Tx, Wx, *_ = ssq_stft(x, padtype='reflect')
Tx, Wx = np.flipud(Tx), np.flipud(Wx)
imshow(Wx, abs=1, ticks=0, 
       title="abs(STFT), f=4 pure sine, 256 samples | padtype='reflect'")
imshow(Tx, abs=1, ticks=0, title="abs(SSQ_STFT)")
#%%
x = x[12:-12]
Tx, Wx, *_ = ssq_stft(x, padtype='reflect')
Tx, Wx = np.flipud(Tx), np.flipud(Wx)
imshow(Wx, abs=1, ticks=0, title="abs(STFT), x=x[12:-12] | padtype='reflect'")
imshow(Tx, abs=1, ticks=0, title="abs(SSQ_STFT)")

#%%
Tx, Wx, *_ = ssq_stft(x, padtype='zero')
Tx, Wx = np.flipud(Tx), np.flipud(Wx)
imshow(Wx, abs=1, ticks=0, title="abs(STFT), x=x[12:-12] | padtype='zero'")
imshow(Tx, abs=1, ticks=0, title="abs(SSQ_STFT)")
```

</details>

![image](https://user-images.githubusercontent.com/16495490/109354511-58320500-7897-11eb-8422-106a8991d45a.png)

**Comment by BakerBunker** [comment]
Any progress on this? The `(len(x) - 1)//hop_length + 1` result is confusing.

**Comment by bmcfee** [comment]
> Any progress on this?

No, nobody has volunteered to work on this.  To reiterate, I don't think there's anything wrong with our implementation and there's no bug to fix - it essentially amounts to a proposed feature request.  My previous comment https://github.com/librosa/librosa/issues/1278#issuecomment-786752819 does precisely lay out what's going on here and how all the parameters interact.
